src/main/java/com/zaxxer/nuprocess/osx/OsxProcess.java

```diff
@@ -103,14 +87,87 @@ public NuProcess start(List<String> command, String[] environment, Path cwd) {
          LibC.posix_spawn_file_actions_destroy(posix_spawn_file_actions);
 
          // After we've spawned, close the unused ends of our pipes (that were dup'd into the child process space)
-         LibC.close(stdinWidow);
-         LibC.close(stdoutWidow);
-         LibC.close(stderrWidow);
+         closePipes();
       }
 
       return this;
    }
 
+    @Override
+   public void run(List<String> command, String[] environment, Path cwd)
+   {
+      callPreStart();
+
+      String[] commands = command.toArray(new String[0]);
+
+      Pointer posix_spawn_file_actions = createPosixPipes();
+      Pointer posix_spawnattr = createPosixSpawnAttributes();
+
+      try {
+         int rc = prepareProcess(environment, cwd, commands, posix_spawn_file_actions, posix_spawnattr);
+
+         if (!checkLaunch()) {
+            return;
+         }
+
+         checkReturnCode(rc, "Invocation of posix_spawn() failed");
+
+         afterStart();
+
+         myProcessor = (IEventProcessor) new ProcessKqueue(this);
+
+         callStart();
+
+         singleProcessContinue();
+
+         myProcessor.run();
+      }
+      catch (RuntimeException e) {
+         // TODO remove from event processor pid map?
+         LOGGER.log(Level.WARNING, "Exception thrown from handler", e);
+         onExit(Integer.MIN_VALUE);
+      }
+      finally {
+         LibC.posix_spawnattr_destroy(posix_spawnattr);
+         LibC.posix_spawn_file_actions_destroy(posix_spawn_file_actions);
+
+         // After we've spawned, close the unused ends of our pipes (that were dup'd into the child process space)
+         closePipes();
+      }
+   }
+
+   private int prepareProcess(String[] environment, Path cwd, String[] commands,
```

Comment:
This is the same logic that was previously in `OsxProcess.start`, refactored into a separate method for reuse in `run` but otherwise unchanged